### ISSUE TITLE
Connection not restored. Help )

### ISSUE THREAD

**Issue by akahan** [nats-io/node-nats#173]
Hi. I use node-nats to connect to the polygon.io. I created a client, connect, subscribe and receive the data. All ok. I also did callbacks on error, on disconnect, on reconnect, on close and so on. Next, I break the network on the router to test reconnecting when network broken. In the end, none of callbacks is not invoked. The connection is not restored. Apparently I'm doing something wrong.

**Comment by aricart** [comment]
Likely in your connect configuration you need to set `maxReconnectAttempts` to `-1` like this:
`var nc = NATS.connect({maxReconnectAttempts: -1});` By default the client will attempt reconnect 10 times, and then call `close()`.

**Comment by aricart** [comment]
Also, this works:
```
'use strict';

var NATS = require('../');

var nc = NATS.connect({maxReconnectAttempts: 3});

nc.on('error', function(e) {
    console.error(e);
    process.exit();
});

nc.on('disconnect', () => console.log('disconnected'));

nc.on('reconnecting', () => console.log('reconnecting'));

nc.on('close', () => console.log('close'));

nc.on('connect', () => console.log('initial connect'));
```

```
albertoricart@barsoom:~/go/src/git...e-nats/examples]$ node connecthandlers.js 
initial connect
disconnected
reconnecting
disconnected
reconnecting
disconnected
reconnecting
disconnected
close
[albertoricart@barsoom:~/go/src/git...e-nats/examples]$ 
```

**Comment by akahan** [comment]
No, it's not working. Weird. 
bash-3.2$ node build/server/FeedFactory/test_polygon.js
initial connect

And then hangs!

**Comment by akahan** [comment]
Yes, when I create a local server, gnatsd, and after connecting the script I kill the server then my script tries to reconnect 3 times, then closes the connection. But why it doesn't work when i break the network?

**Comment by aricart** [comment]
node-nats doesn't have a timer executing the pings (that is a bug) so you are at the mercy of the OS notification which could take hours.

**Comment by aricart** [comment]
As a workaround, you could set-up an interval that sends a ping. Sort of what this does:
https://gist.github.com/aricart/2796171f295fd4bdf59d57a5f2d383e9. However, the real fix needs to come from the library.

**Comment by akahan** [comment]
Many thanks for the prompt reply.

**Comment by akahan** [comment]
Here's what I got:

```
    setup_heartbeat() {
        if (this.heartbeat_timer) return;

        this.logger.info(`Setup heartbeat to ${this.nats.currentServer.url.href}`);

        const heartbeat = () => {
            this.logger.debug(`Ping server ${this.nats.currentServer.url.href}`);

            let timeout = setTimeout(() => {
                try {
                    this.logger.error(`Connection to ${this.nats.currentServer.url.href} lost. Try to reconnecting.`);
                    timeout = null;

                    this.nats.closeStream();
                }
                catch (error) {
                    this.logger.error(error);
                }
            }, 500);

            this.nats.flush(() => {
                if (timeout) {
                    clearTimeout(timeout);
                }
            });
        };

        this.heartbeat_timer = setInterval(heartbeat, HEARTBEAT_INTERVAL);
    }

```

**Comment by aricart** [comment]
the problem with closeStream is fairly aggressive because it effectively closes the nats connection. The solution is really to invoke the code under `stream.on('close')` and have the library re-connect as necessary.

**Comment by akahan** [comment]
No, i try stream.on('close'). So does not work
This causes various errors:
```

events.js:193
      throw er; // Unhandled 'error' event
      ^

Error: connect ETIMEDOUT 172.98.79.116:30403
    at Object._errnoException (util.js:1031:13)
    at _exceptionWithHostPort (util.js:1052:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1195:14)
```
or 
```
events.js:193
      throw er; // Unhandled 'error' event
      ^

Error: connect ENETUNREACH 172.98.79.116:30403
    at Object._errnoException (util.js:1031:13)
    at _exceptionWithHostPort (util.js:1052:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1195:14)
```
or
```
events.js:193
      throw er; // Unhandled 'error' event
      ^

Error: getaddrinfo ENOTFOUND nats2.polygon.io nats2.polygon.io:30402
    at errnoException (dns.js:55:10)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:97:26)
```

**Comment by akahan** [comment]
With closeStream() all work like a charm)

**Comment by aricart** [comment]
This is issue is likely addressed in #174.
